comit/src/network/protocols/orderbook.rs

```diff
@@ -357,4 +452,144 @@ impl Orderbook {
         );
     }
+
+    fn poll(
+        &mut self,
+        _: &mut Context<'_>,
+        _: &mut impl PollParameters,
+    ) -> Poll<
+        NetworkBehaviourAction<
+            EitherOutput<GossipsubRpc, RequestProtocol<TakeOrderCodec>>,
+            BehaviourOutEvent,
+        >,
+    > {
+        if let Some(event) = self.events.pop_front() {
+            return Poll::Ready(NetworkBehaviourAction::GenerateEvent(event));
+        }
+
+        Poll::Pending
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        asset::{self, Erc20, Erc20Quantity},
+        network::test::{await_events_or_timeout, connect, new_swarm},
+    };
+    use atty::{self, Stream};
+    use libp2p::Swarm;
+    use log::LevelFilter;
+    use spectral::prelude::*;
+    use std::str::FromStr;
+    use tracing::{info, subscriber, Level};
+    use tracing_log::LogTracer;
+    use tracing_subscriber::FmtSubscriber;
+
+    #[tokio::test]
+    async fn take_order_request_confirmation() {
+        let comit_level = LevelFilter::Debug;
+        let upstream_level = LevelFilter::Info;
+        init_tracing(comit_level, upstream_level).expect("failed to init tracing");
+
+        // arrange
+
+        let delay = Duration::from_secs(2);
+
+        let (mut alice_swarm, _, _alice_peer_id) = new_swarm(Orderbook::new);
+        let (mut bob_swarm, bob_addr, bob_peer_id) = new_swarm(Orderbook::new);
+
+        // libp2p::Swarm::dial_addr(&mut alice_swarm, bob_addr.clone());
+
+        connect(&mut alice_swarm, &mut bob_swarm).await;
+
+        // TODO: Create helper function for this.
+        let order = Order::new(bob_peer_id.clone(), NewOrder {
+            buy: asset::Bitcoin::from_sat(100),
+            sell: Erc20 {
+                token_contract: Default::default(),
+                quantity: Erc20Quantity::max_value(),
+            },
+            absolute_expiry: 100,
+            maker_addr: bob_addr.clone(),
+        });
+
+        // TODO: Create helper function for this.
+        let bob_refund_identity =
+            crate::bitcoin::Address::from_str("2MufZ6LLCqYTvZnCwzfAjSKn26Xcnr19PBE")
+                .expect("failed to parse bitcoin address");
+        let bob_redeem_identity = identity::Ethereum::random();
+
+        // act
+
+        // Trigger subscription.
+        poll_with_delay(&mut alice_swarm, delay).await;
+        poll_with_delay(&mut bob_swarm, delay).await;
+
+        let _ = bob_swarm
+            .make(order.clone(), bob_refund_identity, bob_redeem_identity)
+            .expect("order id should exist");
+
+        // Trigger publish and receipt of order.
+        poll_with_delay(&mut bob_swarm, delay).await;
+        poll_with_delay(&mut alice_swarm, delay).await;
+
+        let order = alice_swarm
+            .get_orders()
+            .first()
+            .cloned()
+            .expect("Alice has no orders");
+
+        tracing::info!("alice received order");
```

Comment:
No idea.

Reply:
Might be a worthwhile investment to get that working, then you can avoid these checkpoint log statements :)

Reply:
Can you extrapolate on what you mean by 'set breakpoints in your setup' please? My bad for the curt initial response. The only 'breakpoints' I know about are gdb breakpoints and that does not seem relevant here, or am I missing something?

Reply:
I was referring to debugger breakpoints because I saw you writing tracing statements that act as "checkpoints" in your test.

It might be worthwhile getting your setup (emacs + rust-analyzer) to handle breakpoints so you don't have to write these checkpoint trace statements :)